rust/src/mysql/parser.rs

```diff
@@ -0,0 +1,3350 @@
+    },
+    FieldsList {
+        columns: Option<Vec<MysqlColumnDefinition>>,
+    },
+    ResultSet {
+        n_cols: u64,
+        columns: Vec<MysqlColumnDefinition>,
+        eof: MysqlEofPacket,
+        rows: Vec<MysqlResultSetRow>,
+    },
+    BinaryResultSet {
+        n_cols: u64,
+        eof: MysqlEofPacket,
+        rows: Vec<MysqlResultBinarySetRow>,
+    },
+
+    StmtPrepare {
+        statement_id: u32,
+        num_params: u16,
+        params: Option<Vec<MysqlColumnDefinition>>,
+        fields: Option<Vec<MysqlColumnDefinition>>,
+    },
+    StmtFetch,
+}
+
+#[derive(Debug)]
+pub struct MysqlResponse {
+    pub item: MysqlResponsePacket,
+}
+
+#[derive(Debug)]
+pub enum MysqlBEMessage {
+    HandshakeRequest(MysqlHandshakeRequest),
+    Response(MysqlResponse),
+}
+
+#[derive(Debug)]
+pub enum MysqlFEMessage {
+    SSLRequest(MysqlSSLRequest),
+    AuthRequest,
+    Request(MysqlRequest),
+    LocalFileData(u32),
+    HandshakeResponse(MysqlHandshakeResponse),
+}
+
+fn parse_varint(i: &[u8]) -> IResult<&[u8], u64> {
+    let (i, length) = be_u8(i)?;
+    match length {
+        // 251: NULL
+        0xfb => Ok((i, 0)),
+        // 252: value of following 2
+        0xfc => {
+            let (i, v0) = be_u8(i)?;
+            let (i, v1) = be_u8(i)?;
+            let v0 = v0 as u64;
+            let v1 = (v1 as u64) << 8;
+            Ok((i, v0 | v1))
+        }
+        // 253: value of following 3
+        0xfd => {
+            let (i, v0) = be_u8(i)?;
+            let (i, v1) = be_u8(i)?;
+            let (i, v2) = be_u8(i)?;
+            let v0 = v0 as u64;
+            let v1 = (v1 as u64) << 8;
+            let v2 = (v2 as u64) << 16;
+            Ok((i, v0 | v1 | v2))
+        }
+        // 254: value of following 8
+        0xfe => {
+            let (i, v0) = be_u8(i)?;
+            let (i, v1) = be_u8(i)?;
+            let (i, v2) = be_u8(i)?;
+            let (i, v3) = be_u8(i)?;
+            let (i, v4) = be_u8(i)?;
+            let (i, v5) = be_u8(i)?;
+            let (i, v6) = be_u8(i)?;
+            let (i, v7) = be_u8(i)?;
+            let v0 = v0 as u64;
+            let v1 = (v1 as u64) << 8;
+            let v2 = (v2 as u64) << 16;
+            let v3 = (v3 as u64) << 24;
+            let v4 = (v4 as u64) << 32;
+            let v5 = (v5 as u64) << 40;
+            let v6 = (v6 as u64) << 48;
+            let v7 = (v7 as u64) << 56;
+            Ok((i, v0 | v1 | v2 | v3 | v4 | v5 | v6 | v7))
+        }
+        _ => Ok((i, length as u64)),
+    }
+}
+
+pub fn parse_packet_header(i: &[u8]) -> IResult<&[u8], MysqlPacket> {
+    let (i, pkt_len) = le_u24(i)?;
+    let (i, pkt_num) = be_u8(i)?;
+    Ok((i, MysqlPacket { pkt_len, pkt_num }))
+}
+
+fn parse_eof_packet(i: &[u8]) -> IResult<&[u8], MysqlEofPacket> {
+    let (i, _header) = parse_packet_header(i)?;
+    let (i, _tag) = verify(be_u8, |&x| x == 0xfe)(i)?;
+    let (i, warnings) = le_u16(i)?;
+    let (i, status_flags) = le_u16(i)?;
+
```

Comment:
`warnings` and `status_flags` can be parsed only if the following condition is true:
```
if capabilities & CLIENT_PROTOCOL_41 != 0
```
The flags parsed during the handshake phase should be passed here as a parameter to check this condition.

Reply:
When i parse handshake response, i have checked this.

Reply:
Fixed, please check again.